### PR TITLE
introduce mapK operations to concurrent primitives

### PR DESCRIPTION
**Pull request by Odomontois**
Supplying mapK whenever possible, or imapK operations as discussed in #373

### CODE REVIEW

**Comment by Odomontois** [comment]
Note that we can't implement  one-way `mapK` for `Semaphore` because of 
```scala
def withPermit[A](t: F[A]): F[A]
```
Moving this method to mere syntax could potentially make `Semaphore` cleaner and `mapK` possible

**Comment by ChristopherDavenport** [comment]
To implement for Semaphore you need `imapK`

**Comment by Odomontois** [comment]
> To implement for Semaphore you need `imapK`

I didn't understand your comment. 
I did exactly this.

**Comment by SystemFw** [comment]
> I did exactly this.

You did `mapK`, Chris is talking about `imapK` (note the `i`), which requires a bidirectional transformation necessary to translate ops where `F` appears both in negative and positive position (so both covariant and contravariant, hence `invariant`, from where the `i` in `imapK` comes from). 
fs2 Queue has an example.

**Comment by SystemFw** [comment]
ah, my bad, sorry

**Comment by Odomontois** [comment]
> > I did exactly this.
> 
> You did `mapK`, Chris is talking about `imapK` (note the `i`), which requires a bidirectional transformation necessary to translate ops where `F` appears both in negative and positive position (so both covariant and contravariant, hence `invariant`, from where the `i` in `imapK` comes from).
> fs2 Queue has an example.

No. I did `imapK` for Semaphore
https://github.com/typelevel/cats-effect/pull/392/files#diff-60779fbe95602b12729d28efd91cddfcR102

**Comment by SystemFw** [comment]
> Note that we can't implement one-way mapK for Semaphore because of

Just reading this comment, one could parse it as "I couldn't implement mapK", whereas you meant "I couldn't do `mapK`, I needed to do `imapK`". It was clear once I looked at the code again, apologies

File: core/shared/src/main/scala/cats/effect/concurrent/Ref.scala [hunk 1]
```diff
@@ -248,5 +257,22 @@ object Ref {
...
+    override def getAndSet(a: A): G[A] = trans(underlying.getAndSet(a))
+    override def tryUpdate(f: A => A): G[Boolean] = trans(underlying.tryUpdate(f))
+    override def tryModify[B](f: A => (A, B)): G[Option[B]] = trans(underlying.tryModify(f))
+    override def update(f: A => A): G[Unit] = trans(underlying.update(f))
+    override def modify[B](f: A => (A, B)): G[B] = trans(underlying.modify(f))
+    override def tryModifyState[B](state: State[A, B]): G[Option[B]] = trans(underlying.tryModifyState(state))
+    override def modifyState[B](state: State[A, B]): G[B] = trans(underlying.modifyState(state))
+
+    override def access: G[(A, A => G[Boolean])] =
+      trans(F.compose[(A, ?)].compose[A => ?].map(underlying.access)(trans(_)))
```
Comment (Odomontois) [inline core/shared/src/main/scala/cats/effect/concurrent/Ref.scala]: I wonder also if that temporary `Functor[F[(A, A => ?]]` should be cached for performance reasons

**Comment by ChristopherDavenport** [comment]
My apologies at reading comment and not code.

**Comment by Odomontois** [comment]
`f` is minimalistic because its meaning follows from the type, but in the implementation the are new `f` s appearing as arguments of methods.

What naming seems more appropriate to you?

**Pull request merged**